File: 4 Kyu/Python/genNumbers.py
import time
from itertools import combinations_with_replacement

start_time = time.time()

def find_all(sum_dig, digs):
  combinations = combinations_with_replacement([1,2,3,4,5,6,7,8,9], digs)
  nums = list(filter(lambda i: sum(i) == sum_dig, combinations))
  x = lambda i: int(''.join([str(j) for j in i]))
  return [len(nums), x(nums[0]), x(nums[-1])] if len(nums) > 0 else []

print(find_all(35, 9))
print(find_all(10,9))
print(find_all(84,4))
print(find_all(27,3))
print(find_all(35,6))

print("Process finished --- %s seconds ---" % (time.time() - start_time))

# def recur(A, k, out=[], i=0, a = []):
#   if len(out) == k:
#     a.append(out)
#     print(out)
#     return

#   for j in range(i, len(A)):
#     out.append(A[j])
#     recur(A, k, out, j, a)
#     out.pop()

#   return a

# A = [1, 2, 3, 4, 5, 6, 7, 8, 9]
# k = 3
# print(recur(A, k))
# ans = filter(lambda x: sum(x) == 10, a)
#print(a)

File: 4 Kyu/Javascript/genNumbers.js
console.time('log');
const [ nums, sum ] = [ [ 1, 2, 3, 4, 5, 6, 7, 8, 9 ], (n) => n.reduce((a, c) => a + +c, 0) ];

const findAll = (n, r, comb = [], k = 0, res = []) => {
	if (comb.length === r) {
		sum(comb) === n && res.push(Array.from({ length: r }, (_, i) => comb[i]));
		return;
	}
	for (let i = k; i < nums.length; i++) {
		comb.push(nums[i]);
		findAll(n, r, comb, i, res);
		comb.pop();
	}
	return res.length > 0 ? [ res.length, res.slice(0, 1)[0].join(''), res.slice(-1)[0].join('') ] : [];
};

console.log(findAll(10, 3));
console.log(findAll(27, 3));
console.log(findAll(84, 4));
console.log(findAll(35, 6));

console.timeEnd('log');

/* const getPerms = (array, size) => {
	const p = (t, i) => {
		if (t.length === size) {
			result.push(t);
			return;
		}
		if (i + 1 > array.length) return;
		p(t.concat(array[i]), i + 1);
		p(t, i + 1);
	};

	let result = [];
	p([], 0);
	return result;
}; */

/* // Other solution. But its slow!
const sum = (n) => n.reduce((a, c) => a + +c, 0);
const findAll = (n, k, s = '', comb = []) => {
	const [ nums, sum ] = [ [ 1, 2, 3, 4, 5, 6, 7, 8, 9 ], (n) => n.reduce((a, c) => a + +c, 0) ];
	if (s.length >= k) !comb.includes(s.split('').sort().join('')) && sum(s.split('')) === n && comb.push(s);
	else for (let e of nums) findAll(n, k, s + e, comb);
	return comb.length > 0 ? [ comb.length, comb.slice(0, 1)[0], comb.slice(-1)[0] ] : [];
}; */
